sdk/metric/metricdata/metricdatatest/comparisons.go

```diff
@@ -0,0 +1,425 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build go1.18
+// +build go1.18
+
+package metricdatatest // import "go.opentelemetry.io/otel/sdk/metric/metricdata/metricdatatest"
+
+import (
+	"bytes"
+	"fmt"
+	"reflect"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+)
+
+// equalResourceMetrics returns true when ResourceMetrics are equal. It
+// returns false when they differ, along with the reasons why they differ.
+//
+// The ScopeMetrics each ResourceMetrics contains are compared based on
+// containing the same ScopeMetrics, not the order they are stored in.
+func equalResourceMetrics(a, b metricdata.ResourceMetrics) (equal bool, reasons []string) {
+	equal = true
+	if !a.Resource.Equal(b.Resource) {
+		equal, reasons = false, append(reasons, notEqualStr("Resources", a.Resource, b.Resource))
+	}
+
+	var r string
+	equal, r = compareDiff(diffSlices(
+		a.ScopeMetrics,
+		b.ScopeMetrics,
+		func(a, b metricdata.ScopeMetrics) bool {
+			equal, _ := equalScopeMetrics(a, b)
+			return equal
+		},
+	))
+	if !equal {
+		reasons = append(reasons, fmt.Sprintf("ResourceMetrics ScopeMetrics not equal:\n%s", r))
+	}
+	return equal, reasons
+}
+
+// equalScopeMetrics returns true when ScopeMetrics are equal. It returns
+// false when they differ, along with the reasons why they differ.
+//
+// The Metrics each ScopeMetrics contains are compared based on containing the
+// same Metrics, not the order they are stored in.
+func equalScopeMetrics(a, b metricdata.ScopeMetrics) (equal bool, reasons []string) {
+	equal = true
+	if a.Scope != b.Scope {
+		equal, reasons = false, append(reasons, notEqualStr("Scope", a.Scope, b.Scope))
+	}
+
+	var r string
+	equal, r = compareDiff(diffSlices(
+		a.Metrics,
+		b.Metrics,
+		func(a, b metricdata.Metrics) bool {
+			equal, _ := equalMetrics(a, b)
+			return equal
+		},
+	))
+	if !equal {
+		reasons = append(reasons, fmt.Sprintf("ScopeMetrics Metrics not equal:\n%s", r))
+	}
+	return equal, reasons
+}
+
+// equalMetrics returns true when Metrics are equal. It returns false when
+// they differ, along with the reasons why they differ.
+func equalMetrics(a, b metricdata.Metrics) (equal bool, reasons []string) {
+	equal = true
+	if a.Name != b.Name {
+		equal, reasons = false, append(reasons, notEqualStr("Name", a.Name, b.Name))
+	}
+	if a.Description != b.Description {
+		equal, reasons = false, append(reasons, notEqualStr("Description", a.Description, b.Description))
+	}
+	if a.Unit != b.Unit {
+		equal, reasons = false, append(reasons, notEqualStr("Unit", a.Unit, b.Unit))
+	}
+
+	var r []string
+	equal, r = equalAggregations(a.Data, b.Data)
```

Comment:
We are overwriting `equal` in all of these cases.  if the aggregations are the same but the name is different then this will return true.

Reply:
Resolved now that equalAggregations no longer returns a bool.